### PR TITLE
Optimize `decide_worker`

### PR DESCRIPTION
**Pull request by jakirkham**
Annotate variables in `decide_worker` (both method and function). Deduplicate some code within the `decide_worker` method. Also tune the `decide_worker` function fast path.

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -4667,36 +4668,31 @@ def decide_worker(self, ts: TaskState):
...
-                worker,
+            assert ws is None or isinstance(ws, WorkerState), (
+                type(ws),
+                ws,
             )
-            assert worker.address in self.workers
+            assert ws.address in self.workers
 
-        return worker
+        return ws
```
Comment (mrocklin) [inline distributed/scheduler.py:4695]: Thank you for standardizing pronouns along the way

**Comment by mrocklin** [comment]
It looks like we're going the route of doing easy optimizations before splitting things out to a cythonized class and a non-cythonized class.  Is this correct?

File: distributed/scheduler.py (hunk 1 above)
Comment (jakirkham) [reply distributed/scheduler.py:4695]: It's certainly made it a lot easier to understand where things can be annotated 🙂

**Comment by jakirkham** [comment]
Yes and no. I started a refactoring branch as well. Though I think the particular transition these changes are associated with, `transition_waiting_processing`, is probably the hardest to refactor in a meaningful way atm. It also happens to be the 2nd slowest transition. So wanted to make a concerted effort to unblock performance for this transition before going ahead with the refactor.

**Pull request merged**